### ISSUE TITLE
GPU not working

### ISSUE THREAD

**Issue by road2react** [mkiol/dsnote#32]
Selecting GPU to transcribe an audio file is causing a crash

```
QIBusPlatformInputContext: invalid portal bus.
QSocketNotifier: Can only be used with threads started with QThread
qt.qpa.qgnomeplatform: Could not find color scheme  ""
whisper_init_from_file_no_state: loading model from '/home/user/.var/app/net.mkiol.SpeechNote/cache/net.mkiol/dsnote/speech-models/en_whisper_small.ggml'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51864
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 768
whisper_model_load: n_audio_head  = 12
whisper_model_load: n_audio_layer = 12
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 768
whisper_model_load: n_text_head   = 12
whisper_model_load: n_text_layer  = 12
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 9
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 3
whisper_model_load: mem required  =  459.00 MB (+   16.00 MB per decoder)
whisper_model_load: adding 1607 extra tokens
whisper_model_load: model ctx     =  180.95 MB
ggml_opencl: selecting platform: 'Clover'
ggml_opencl: selecting device: 'AMD Radeon RX 6800M (navi22, LLVM 15.0.7, DRM 3.54, 6.5.5-1-linux)'
ggml_opencl: device FP16 support: false
ggml_opencl: kernel compile error:

fatal error: cannot open file '/usr/lib/x86_64-linux-gnu/GL/default/share/clc/gfx1031-amdgcn-mesa-mesa3d.bc': No such file or directory
```

**Comment by mkiol** [comment]
It is unfortunate but right now GPU acceleration for Whisper works only for older AMD cards. The cause of this problem is in Flatpak packaging and isolation Flatpak provides.

GPU acceleration for AMD uses OpenCL and OpenCL provider shiped with Flatpak runtime is "Clover" (a part of Mesa). Clover doesn't support new AMD chips. This is a reason of the crash you observed.

There is also another implementation directly from AMD "ROCm-OpenCL". It works fine on new GPUs. The problem is that Flatpak runtime doesn't have it. I've tried to ship ROCm with Speech Note package but it is too heavy (more than 5GB additional size). 

I'm investigating and looking for any workaround for this problem.

**Comment by mkiol** [comment]
Good news. I think I've managed to overcome this problem. The solution is not super straight forward but seems to work well.

Could you please install the following test version:

```
flatpak install --user https://dl.flathub.org/build-repo/52052/net.mkiol.SpeechNote.flatpakref
```

Due to flatpak sandboxing, app by default is not able to access AMD rocr driver. To fix this, you need to override permissions with following command.

```
flatpak override --user --device=all net.mkiol.SpeechNote
```

than, run the test app (`--verbose` is for debug logs)

```
flatpak run net.mkiol.SpeechNote//test --verbose
```

Enable in the Setting option 'Use GPU acceleration for Whisper' and check what devices are listed in 'GPU device'. Make sure that something like 'OpenCL, AMD Accelerated Parallel Processing, ...' is selected. It should be when 'Auto' is selected.

Try speech to text with Whisper.

Terminal should be full of logs, please find (this is shown only when Whisper is used for the first time after app restart)

```
ggml_opencl: selecting platform: ' AMD Accelerated Parallel Processing'
ggml_opencl: selecting device: 'gfx.....'
```

If you see this lines, it means that GPU was used for TTS.

Measurements for similar audio sample with Whisper Large on my system with Radeon RX 6950 XT:
- CPU: processing time  7184ms
- GPU: processing time  4366ms

Not huge but significant improvement. 

I would be super grateful for your help.

**Comment by mkiol** [comment]
The fix in included in the latest 4.2.1 release.

4b342ac46972f628ec8b9b208732d881c78c159d
8259fa7ba19e5bec97f2f82176ac73205ddcae70

I hope the problem is resolved. If not, please re-open the issue.

Thanks

**Comment by road2react** [comment]
I'm now getting a different issue:

After opening the file, this shows up in logs:

```
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51864
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 768
whisper_model_load: n_audio_head  = 12
whisper_model_load: n_audio_layer = 12
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 768
whisper_model_load: n_text_head   = 12
whisper_model_load: n_text_layer  = 12
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 9
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 3
whisper_model_load: adding 1607 extra tokens
whisper_model_load: model ctx     =  180.95 MB
ggml_opencl: selecting platform: 'AMD Accelerated Parallel Processing'
ggml_opencl: selecting device: 'gfx1031'
ggml_opencl: device FP16 support: true
whisper_model_load: model size    =  180.71 MB
whisper_init_state: kv self size  =   15.75 MB
whisper_init_state: kv cross size =   52.73 MB
whisper_init_state: compute buffer (conv)   =   19.96 MB
whisper_init_state: compute buffer (encode) =  122.04 MB
whisper_init_state: compute buffer (cross)  =    5.86 MB
whisper_init_state: compute buffer (decode) =   36.17 MB
```

The progress gets stuck at 100% and there is no text in the large output textbox.

![image](https://github.com/mkiol/dsnote/assets/93143714/292bb02a-f450-484f-8481-bc44662fb036)

Edit: this happens when using cpu as well

Edit 2: it works with FLAC files, the issue is happening on ogg opus files. Will test more formats

**Comment by mkiol** [comment]
Thanks for the update.

Could you please make a test with `--verbose` option?

This "Transcribing audio file... 100%" is a little bit misleading. Actually "100%" means that 100% of audio data was pushed to the STT decoder. When using Whisper, the decoding part might take the majority of time, so this progress indication is far from perfection unfortunately :/

**Comment by mkiol** [comment]
Ok, I can reproduce it. This "Transcribing audio file... 100%" is caused by buggy handling of unsupported codec. Right now Opus is not supported. 

Thanks for catching it 👍🏿

**Comment by mkiol** [comment]
This is not related to GPU, so closing.

"Opus" issue is already reported in https://github.com/mkiol/dsnote/issues/45.

**Comment by kmoad** [comment]
### TLDR

I found the solution. See edit below. Leaving this up in case it's helpful.

## Original Issue

Using flatpak v4.3.0 and RX6800 GPU with device=all permissions on.

In the GPU acceleration settings I have these two device options.

```
OpenCL, AMD Accelerated Parallel Processing, gfx1030
OpenCL, Clover, AMD Radeon RX 6800 (radeonsi, navi21, LLVM 15.0.7, DRM 3.54, 6.6.7-200.fc39.x86_64)
```

When doing speech-to-text with FasterWhisper Large-v2, neither option moves the processing to the GPU.

Running with --verbose gets these log lines:

```
[D] 01:51:12.838 0x7f70fee17d00 add_hip_devices:266 - scanning for hip devices
[W] 01:51:12.838 0x7f70fee17d00 hip_api:170 - failed to open hip lib: libamdhip64.so: cannot open shared object file: No such file or directory
[D] 01:51:12.838 0x7f70fee17d00 add_opencl_devices:303 - scanning for opencl devices
[D] 01:51:12.890 0x7f70fee17d00 add_opencl_devices:320 - opencl number of platforms: 2
[D] 01:51:12.890 0x7f70fee17d00 add_opencl_devices:345 - opencl platform: 0, name=AMD Accelerated Parallel Processing, vendor=Advanced Micro Devices, Inc.
[D] 01:51:12.890 0x7f70fee17d00 add_opencl_devices:359 - opencl number of devices: 1
[D] 01:51:12.890 0x7f70fee17d00 add_opencl_devices:383 - opencl device: 0, platform name=AMD Accelerated Parallel Processing, device name=gfx1030, types=[GPU, ]
[D] 01:51:12.890 0x7f70fee17d00 add_opencl_devices:345 - opencl platform: 1, name=Clover, vendor=Mesa
[D] 01:51:12.890 0x7f70fee17d00 add_opencl_devices:359 - opencl number of devices: 1
[D] 01:51:12.890 0x7f70fee17d00 add_opencl_devices:383 - opencl device: 0, platform name=Clover, device name=AMD Radeon RX 6800 (radeonsi, navi21, LLVM 15.0.7, DRM 3.54, 6.6.7-200.fc39.x86_64), types=[GPU, ]

```

Are newer AMD GPUs supported in 4.3.0? 

I see that main branch `flatpak/` dir there are build files for amd and nvidia extensions. Are these available to be installed, or for a future version? I'm fairly un-advanced with flatpak.

Things work great out-of-the-box on my Nvidia system though. Which is wonderful and not easy to get right. So thanks for that!

### Edit

I built the flatpak and the amd extension from main. I now have "ROCm, 0, AMD Radeon RX 6800" as the only acceleration option. GPU still not used by "FasterWhisper Large-v2", but is used by "Whisper Large-v3".

After switching back to public v4.3.0 the acceleration option "OpenCL, AMD Accelerated Parallel Processing, gfx 1030" is available. This may have been from fiddling with the "Graphics Card Support" checkboxes in the Other tab of settings.

The "Whisper Large-v3" model does use GPU. Though slower than the upcoming ROCm option.

So it appears my issue was with model selection, not Speech Note. I'm looking forward to the ROCm extension release, it's very fast.

**Comment by mkiol** [comment]
@kmoad

Thank you for testing new "modular" flatpak package. I'm impressed that you managed to install everything! This is actually pre-beta version.

Right now, Faster Whisper supports only NVIDIA CUDA acceleration. This is a limitation of [CTranslate2](https://github.com/OpenNMT/CTranslate2) library which is behind the STT processing. My fault I didn't communicate this in the app. I the next version, there will be "Model info" section where you will be able to check all the properties of the model, including information about possible GPU acceleration.

As you already noticed, new version brings support for ROCm and ROCm is veeery efficient with [whisper.cpp](https://github.com/ggerganov/whisper.cpp) models (**it might be confusing, that in the app "Whisper" are actually "whisper.cpp" models**) . I didn't make any benchmark yet but I think newest wisper.cpp provides similar speed as Faster Whisper. If you are on AMD, whisper.cpp (=Whisper) will be best for you.

If it is not a problem, would you be able to check also ROCm with any TTS Coqui model? To do so, please enable "Use GPU acceleration in "Setting->Text to Speech". If it crash, enable also "Settings->Other->Override GPU version". I will be grateful for your help! Thanks.

**Comment by kmoad** [comment]
Tested 3889b39 with a few different Coqui voices

- `Coqui Jenny Female`: Used GPU. Sounded good. Seemed bottlenecked by GPU processing. It read one sentence as a time, and GPU was at 100% between the sentences.
- `Coqui VCTK VITS P376 Male`: Used GPU. Sounded ok. Read continuously.
- `Coqui STTX-v2.0.2` and `Coqui STTX-v2.0.3`. Did not work. The "Read" button was disabled. Toggling "Override GPU version" did not change anything. Some possibly relevant logs below.

```
[D] 17:22:41.633 0x7effecd82d00 () - tts models changed
[D] 17:22:41.633 0x7effecd82d00 () - update listen
[W] 17:22:41.634 0x7effecd82d00 () - no available tts models for in mnt
[W] 17:22:41.634 0x7effecd82d00 () - no available tts models for out mnt
[D] 17:22:41.634 0x7effecd82d00 () - ttt models changed
[D] 17:22:41.637 0x7effecd82d00 () - mnt langs changed
[D] 17:22:41.637 0x7effecd82d00 () - update listen
[W] 17:22:41.637 0x7effecd82d00 () - no available mnt langs
[W] 17:22:41.637 0x7effecd82d00 () - no available mnt out langs

** (dsnote:2): WARNING **: 17:23:04.987: atk-bridge: get_device_events_reply: unknown signature
[D] 17:23:37.770 0x7effecd82d00 () - app active tts model: "en_coqui_xtts202" => "en_coqui_jenny_jenny"
[D] 17:23:37.770 0x7effecd82d00 () - update listen
[D] 17:23:40.459 0x7effecd82d00 () - app active tts model: "en_coqui_jenny_jenny" => "en_coqui_vctk_vits_p376"
[D] 17:23:40.459 0x7effecd82d00 () - update listen
[D] 17:23:41.559 0x7effecd82d00 () - app active tts model: "en_coqui_vctk_vits_p376" => "en_coqui_xtts202"
[D] 17:23:41.559 0x7effecd82d00 () - update listen
[D] 17:23:42.632 0x7effecd82d00 () - app active tts model: "en_coqui_xtts202" => "en_coqui_xtts203"
[D] 17:23:42.632 0x7effecd82d00 () - update listen
```

Happy to keep testing!

**Comment by mkiol** [comment]
@kmoad Many many thanks!

> Coqui STTX-v2.0.2 and Coqui STTX-v2.0.3

These are "Voice Cloning" models. To make them work, you need to record a voice sample. You can do it in "Voice samples" menu in the main app toolbar.

![image](https://github.com/mkiol/dsnote/assets/1420902/d714ad9a-a599-40fd-9fde-a67042d13624)

I would be very grateful you could make a test after syncing to the latest commit. I just pushed few bug fixes.

Thanks!

**Comment by kmoad** [comment]
Tested e28a78d. Coqui XTTS worked with ROCm. The GPU processing speed mostly kept up with sentence-by-sentence transcription, but there were pauses with some short sentences. This may not be something SpeechNote can fix though.

There was a crash though. After transcribing 40ish sentences, I got `Too many open files` errors. Here's the full log.

[dsnote_verbose.txt](https://github.com/mkiol/dsnote/files/13789282/dsnote_verbose.txt)

Last, and kinda personal preference. When I don't have the SpeechNote window active, there are notifications for each sentence that is being transcribed. I know I can turn them off entirely, but per-sentence seems to often.

**Comment by mkiol** [comment]
> There was a crash though. After transcribing 40ish sentences, I got Too many open files errors. Here's the full log.

Nice catch! Thank you for all your tests.

> Last, and kinda personal preference. When I don't have the SpeechNote window active, there are notifications for each sentence that is being transcribed. I know I can turn them off entirely, but per-sentence seems to often.

Got it. I will try to improve in in the next release.
